src/Storages/StorageFuzzJSON.cpp

```diff
@@ -0,0 +1,695 @@
+
+    static Type getType(const JSONValue & v);
+
+    // The node value must be one of the following:
+    // Examples: 5, true, "abc"
+    std::optional<Field> fixed;
+    // Examples: [], ["a"], [1, true]
+    std::optional<JSONNodeList> array;
+    // Examples: {}, {"a": [1,2], "b": "c"}
+    std::optional<JSONNodeList> object;
+};
+
+JSONValue::Type JSONValue::getType(const JSONValue & v)
+{
+    if (v.fixed)
+    {
+        assert(!v.array);
+        assert(!v.object);
+        return JSONValue::Type::Fixed;
+    }
+    else if (v.array)
+    {
+        assert(!v.fixed);
+        assert(!v.object);
+        return JSONValue::Type::Array;
+    }
+    else if (v.object)
+    {
+        assert(!v.fixed);
+        assert(!v.array);
+        return JSONValue::Type::Object;
+    }
+    else
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Failed to determine JSON node type.");
+}
+
+// A node represents either a JSON field (a key-value pair) or a JSON value.
+// The key is not set for the JSON root and for the array items.
+struct JSONNode
+{
+    std::optional<String> key;
+    JSONValue value;
+};
+
+#if USE_SIMDJSON
+using ParserImpl = DB::SimdJSONParser;
+#elif USE_RAPIDJSON
+using ParserImpl = DB::RapidJSONParser;
+#endif
+
+std::optional<Field> getFixedValue(const ParserImpl::Element & e)
+{
+    return e.isBool()  ? e.getBool()
+        : e.isInt64()  ? e.getInt64()
+        : e.isUInt64() ? e.getUInt64()
+        : e.isDouble() ? e.getDouble()
+        : e.isString() ? e.getString()
+        : e.isNull()   ? Field()
+                       : std::optional<Field>();
+}
+
+void traverse(const ParserImpl::Element & e, std::shared_ptr<JSONNode> node)
+{
+    checkStackSize();
+
+    assert(node);
+
+    auto & val = node->value;
+    if (e.isObject())
+    {
+        const auto & obj = e.getObject();
+        if (!val.object)
+            val.object = JSONNodeList{};
+
+        for (const auto [k, v] : obj)
+        {
+            auto child = std::make_shared<JSONNode>();
+            child->key = k;
+            traverse(v, child);
+            val.object->push_back(child);
+        }
+    }
+    else if (e.isArray())
+    {
+        if (!val.array)
+            val.array = JSONNodeList{};
+
+        const auto arr = e.getArray();
+        for (const auto a : arr)
+        {
+            auto child = std::make_shared<JSONNode>();
+            traverse(a, child);
+            val.array->push_back(child);
+        }
+    }
+    else
+    {
+        auto field = getFixedValue(e);
+        if (!field)
+            throw Exception(ErrorCodes::INCORRECT_DATA, "Failed to parse a fixed JSON value.");
+
+        val.fixed = std::move(field);
+    }
+}
+
+std::shared_ptr<JSONNode> parseJSON(const String & json)
+{
+    std::string_view view{json.begin(), json.end()};
+    ParserImpl::Element document;
+    ParserImpl p;
+
+    if (!p.parse(json, document))
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Failed to parse JSON string.");
```

Comment:
These exception messages are not really informative, consider adding more context to simplify debugging.

(found by `00002_log_and_exception_messages_formatting`: [link](https://s3.amazonaws.com/clickhouse-test-reports/57339/84abbafcc5895d21fde4417b541a43c33cd00dcf/stateless_tests__msan__[5_6].html))